Adds 'Define element' functionality

YuriyYatsyk opened the pull request:
Similarly to 'Run Selection', 'Define On' menus that are applicable only for top-level functions & take into account namespaces.

'Define xxx On' is also available in the Structure View.

'Define On' shortcuts - Ctrl+Alt+R to redefine, Ctrl+Alt+Shift+R to force server selection popup.
'Evaluate Selection On' shortcuts changed to Ctrl+Alt+E and Ctrl+Alt+Shift+E

a2ndrade commented:
@YuriyYatsyk thanks for the PR, could you address a couple of things I found?:

1) In `KServerProcessHandler#execute`, `String s = new String((char[])conn.k("{.Q.s value x}", q.toCharArray()))` is breaking for me b/c even though you're evaluating an expression that always yields a string, the request handler (in q) can be overridden to return structured data (or any shape). So, for me, this line is always breaking with `Ljava.lang.Object; cannot be cast to [C` (i.e. `Object[]` cannot be cast to `char[]`

2) Unchecking the TLS box doesn't seem to update the config (somewhere). That is, even after I unchecked the checkbox, it still kept trying to use TLS. This is what I did:
a) enabled TLS checkbox, made request
b) q logs a `'TLS not enabled` message (and rejects the request)
c) a `Connection reset` message is printed to the console in intellij
d) unchecked the box, made request
e) kept getting `TLS not enabled` message

Thanks!

YuriyYatsyk commented:
How would you fix 1? If custom q request handler does not honor the contract and does not return the value of the expression it evaluated, what would you do with the result? Show toString()?

> On Sep 12, 2019, at 11:28 PM, Antonio Andrade <notifications@github.com> wrote:
> 
> @YuriyYatsyk thanks for the PR, could you address a couple of things I found?:
> 
> In KServerProcessHandler#execute, String s = new String((char[])conn.k("{.Q.s value x}", q.toCharArray())) is breaking for me b/c even though you're evaluating an expression that always yields a string, the request handler (in q) can be overridden to return structured data (or any shape). So, for me, this line is always breaking with Ljava.lang.Object; cannot be cast to [C (i.e. Object[] cannot be cast to char[]
> 
> Unchecking the TLS box doesn't seem to update the config (somewhere). That is, even after I unchecked the checkbox, it still kept trying to use TLS. This is what I did:
> a) enabled TLS checkbox, made request
> b) q logs a 'TLS not enabled message (and rejects the request)
> c) a Connection reset message is printed to the console in intellij
> 
> Thanks!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

a2ndrade commented:
Could you check if the result is an instance of `Object[]` and, if it's, call `Arrays.deepToString(...)`, otherwise, just call `toString()` as you suggested.

YuriyYatsyk commented:
Ok

> On Sep 13, 2019, at 4:57 PM, Antonio Andrade <notifications@github.com> wrote:
> 
> Could you check if the result is an instance of Object[] and, if it's, call Arrays.deepToString(...), otherwise, just call toString() as you suggested.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

YuriyYatsyk commented:
@a2ndrade Pushed fixes for both.

For 2 TLS flag is now part of unique server identifier, hence new connection will be open if the flag changes. This approach allows not worrying about what to do if someone, e.g., deletes a server from the settings while the connection is still open.

a2ndrade commented on src/main/java/com/appian/intellij/k/actions/KServerProcessHandler.java:
```diff
@@ -90,6 +92,58 @@ void execute(ConsoleView console, String q) {
       closeConnection();
+    } finally {
+      if (console instanceof ConsoleViewImpl) {
+        ((ConsoleViewImpl)console).requestScrollingToEnd();
```
@YuriyYatsyk added this to make sure console scrolls to the end when executing actions. Otherwise, it's hard to know if something happened (unless the user manually scrolls).

a2ndrade commented:
I did some quick testing and it doesn't seem the server specs are being properly updated. Similar to the TLS config, the password is not correctly updated. How to reproduce:
1. Right click, select "Definte ... on", "New Server..."
2. Configure server *except* for the password
3. Run. We get `access` an error (expected)
4. Edit the server configuration to add a password
5. Rerun accion, we still get an `access` error.

a2ndrade commented:
The keyboard shortcuts names show the implementation class rather than a friendly name (see attachment). And there are 2 "Separator" rows (not sure where that's coming from)

<img width="737" alt="Screen Shot 2019-09-15 at 3 37 07 PM" src="https://user-images.githubusercontent.com/360167/64926713-b61b7b80-d7ce-11e9-81c2-d701963a955a.png">

Thanks!

YuriyYatsyk commented:
Once the console is open, it has it’s own server that does not get updated. Closing console fixes the issue, but I see how that can become confusing. Should be fixable.

> On Sep 15, 2019, at 3:36 PM, Antonio Andrade <notifications@github.com> wrote:
> 
> I did some quick testing and it doesn't seem the server specs are being properly updated. Similar to the TLS config, the password is not correctly updated. How to reproduce:
> 
> Right click, select "Definte ... on", "New Server..."
> Configure server except for the password
> Run. We get access an error (expected)
> Edit the server configuration to add a password
> Rerun accion, we still get an access error.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

YuriyYatsyk commented:
I’ll take care of separators, but I don’t have a Mac and shortcuts seem ok for me, not sure how to reproduce that.

> On Sep 15, 2019, at 3:39 PM, Antonio Andrade <notifications@github.com> wrote:
> 
> The keyboard shortcuts names show the implementation class rather than a friendly name (see attachment). And there are 2 "separator" rows (not sure where that's coming from)
> 
> 
> 
> Thanks!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

a2ndrade commented:
Thanks, about the shortcuts.. are you saying that in windows the Keymap configuration (shown in the attachment) reads "`Run selection popup`" instead of "`KDefineElementOnActiveServerAction`"?

Actually, I just realized the `<action/>` elements in `plugin.xml` have both a `text` and `description` attributes and that you explicitly set the text to be `KDefineElementOnActiveServerAction` and the description to be `Run selection popup`. I think the `text` attribute should also have a friendly name e.g. `Run selection popup`. Any reason why you chose the class name instead?

YuriyYatsyk commented:
Yes, I saw that after I replied. Windows works the same way as Mac, I misinterpreted your observation.

> On Sep 15, 2019, at 10:45 PM, Antonio Andrade <notifications@github.com> wrote:
> 
> Thanks, about the shortcuts.. are you saying that in windows the Keymap configuration (shown in the attachment) reads "Run selection popup" instead of "KDefineElementOnActiveServerAction"?
> 
> Actually, I just realized the <action/> elements in plugin.xml have both a text and description attributes and that you explicitly set the text to be KDefineElementOnActiveServerAction and the description to be Run selection popup. I think the text attribute should also have a friendly name e.g. Run selection popup. Any reason why you chose the class name instead?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

YuriyYatsyk commented:
@a2ndrade 

Pushed fixes:
1. TLS/password gets picked up correctly if connection cannot be established
2. User friendly names in the keymap
3. No more 'Separator' in the keymap

I wanted to add a separator to popup menu after 'Define xxx On' submenu, the only way I could do that is by adding an action after action group using Separator class as action implementation. That works fine, but it adds Separator item to the keymap. I removed these not to pollute the keymap, but that also removed separators from the menu. It would be much nicer if menus did have separators - any ideas how to add them?

a2ndrade commented:
@YuriyYatsyk I made a couple of minor changes. I changed the key bindings so the shortcuts don't collide with built-in OSX shortcuts, could you try them on Windows?

Also, for now, I'd like to recreate the q connection on every action to make sure the latest config is always picked up (which was not happening when password, for example, was removed while the tab/window was opened).

Let me know if the shortcuts work for you so I can merge and kick off another release. Thanks!

YuriyYatsyk commented:
No, shortcuts no longer work on windows

a2ndrade commented:
@YuriyYatsyk I changed the modifier key from `ctl` to `meta`. In windows, the `meta` key is mapped to the `Windows` key, [I believe](https://www.computerhope.com/jargon/m/meta-key.htm). Could you try using the `Windows` key? e.g. `windows + alt + D`

YuriyYatsyk commented:
@a2ndrade  Yep, I saw that and tried it - nope, windows key does not work like meta.

YuriyYatsyk commented:
@a2ndrade In fact, `Windows+Alt+R` simply opens `Refactor` menu.

a2ndrade commented:
@YuriyYatsyk  Updated the shortcuts (back to using `ctrl`) but didn't use `ctrl alt D` because it collides with the built-in OSX shortcut for "Show or hide the Dock" (see [Option-Command-D](https://support.apple.com/en-us/HT201236)) Are you OK with the changes?

YuriyYatsyk commented:
@a2ndrade Yes, thanks, seems to work now.
